File: src/main/java/controller/KafkaBrokerConfigMonitor.java
package controller;

import com.google.common.base.Joiner;
import io.javaoperatorsdk.operator.OperatorException;
import io.javaoperatorsdk.operator.processing.LifecycleAware;
import org.apache.curator.framework.CuratorFramework;
import org.apache.curator.framework.recipes.cache.ChildData;
import org.apache.curator.framework.recipes.cache.TreeCache;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.api.libs.json.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class KafkaBrokerConfigMonitor implements LifecycleAware {
    private final Logger logger = LoggerFactory.getLogger(KafkaBrokerConfigMonitor.class);
    private final Map<String, TreeCache> clusterToZNodeMap = new HashMap<>();
    private final String kafkaBasePath;

    public KafkaBrokerConfigMonitor(CuratorFramework curator, String kafkaBasePath, List<String> kafkaClusters) {
        this.kafkaBasePath = kafkaBasePath;
        kafkaClusters.forEach(
                cluster -> clusterToZNodeMap.put(cluster, new TreeCache(curator, kafkaBasePath +"/" + cluster))
        );
    }

    public Map<String, String> getClusterIps() {
        HashMap<String, List<String>> map = new HashMap<>();
        clusterToZNodeMap.forEach((clusterPrefix, treeCache) -> {
            Map<String, ChildData> brokerNodeAndDataMap =
                    treeCache.getCurrentChildren(kafkaBasePath + "/" + clusterPrefix);
            if (brokerNodeAndDataMap != null && !brokerNodeAndDataMap.isEmpty()) {
                brokerNodeAndDataMap.values().forEach(childData -> {
                    try {
                        byte[] brokerData = childData.getData();
                        String data = new String(brokerData);

                        logger.info("ChildPath: " + childData.getPath());
                        logger.error("Broker data: " + data);
                        parseAndPopulateMap(map, data);
                    } catch (Exception ex) {
                        logger.error("Cluster IP Exception ex: " + ex);
                    }
                });
            }
        });
        return map.entrySet().stream().collect(
                Collectors.toMap(
                        Map.Entry::getKey,
                        e -> Joiner.on(",").join(e.getValue())
                ));
    }

    private static void parseAndPopulateMap(HashMap<String, List<String>> map, String data) {
        JsObject jsValue = (JsObject) Json.parse(data);
        JsArray nodeAddresses = (JsArray) jsValue.underlying().get("nodeAddresses").get();

        String port = ((JsNumber) jsValue.underlying().get("port").get()).value().toString();

        JsString cluster = (JsString) ((JsObject) jsValue.underlying().get("properties").get())
                .underlying().get("node.cluster").get();

        map.putIfAbsent(cluster.value(), new ArrayList<>());
        List<String> ipsForCluster = map.get(cluster.value());

        nodeAddresses.value()
                .toStream()
                .map(x -> ((JsString) x).value() + ":" + port)
                .foreach(ipsForCluster::add);
    }

    @Override
    public void start() throws OperatorException {
        try {
            for (TreeCache treeCache : clusterToZNodeMap.values()) {
                treeCache.start();
            }
        } catch (Exception e) {
            logger.error("Monitor Failed to Start",  e);
            e.printStackTrace();
        }
    }

    @Override
    public void stop() throws OperatorException {
        for (TreeCache treeCache : clusterToZNodeMap.values()) {
            treeCache.close();
        }
    }
}

File: src/main/scala/controller/KafkaBrokerConfigMonitor.scala
package controller

import io.javaoperatorsdk.operator.processing.LifecycleAware
import org.apache.curator.framework.CuratorFramework
import org.apache.curator.framework.recipes.cache.{ChildData, TreeCache}
import org.slf4j.{Logger, LoggerFactory}
import play.api.libs.json._

import scala.collection.mutable
import scala.collection.mutable.ListBuffer


class KafkaBrokerConfigMonitor extends LifecycleAware {

  private val logger: Logger = LoggerFactory.getLogger(classOf[KafkaBrokerConfigMonitor])
  private val clusterToZNodeMap = new mutable.HashMap[String, TreeCache]
  private var kafkaBasePath: String = null

  def this(curator: CuratorFramework, kafkaBasePath: String, kafkaClusters: List[String]) = {
    this()
    this.kafkaBasePath = kafkaBasePath
    kafkaClusters.foreach((cluster: String) => clusterToZNodeMap += (cluster -> new TreeCache(curator, kafkaBasePath + "/" + cluster)))
  }

  override def start(): Unit = {
    try {
      for ((_, treeCache) <- clusterToZNodeMap)
        treeCache.start()
    }
    catch {
      case e: Exception =>
        logger.error("Monitor Failed to Start", e)
        e.printStackTrace()
    }
  }

  override def stop(): Unit = {
    for ((_, treeCache) <- clusterToZNodeMap)
      treeCache.close()
  }

  private def parseAndPopulateMap(map: mutable.HashMap[String, List[String]], data: String): Unit = {
    val jsValue = Json.parse(data).asInstanceOf[JsObject]
    val nodeAddresses = jsValue.value("nodeAddresses").asInstanceOf[JsArray]
    val port = jsValue.value("port").asInstanceOf[JsNumber].value.toString
    val cluster = jsValue.value("properties").asInstanceOf[JsObject].value("node.cluster").asInstanceOf[JsString]
    val ipsForClusterBuffer = new ListBuffer[String]()
    map.getOrElse(cluster.value, new ListBuffer[String].toList).foreach((ele) => ipsForClusterBuffer.addOne(ele))
    nodeAddresses.value.to(LazyList).map((x: JsValue) => x.asInstanceOf[JsString].value + ":" + port).foreach((ele) => {
      ipsForClusterBuffer += ele
    })
    map.put(cluster.value, ipsForClusterBuffer.toList)
  }

  def getClusterIps: mutable.Map[String, String] = {
    val map = new mutable.HashMap[String, List[String]]
    for ((clusterPrefix, treeCache) <- clusterToZNodeMap) {
      val brokerNodeAndDataMap = treeCache.getCurrentChildren(kafkaBasePath + "/" + clusterPrefix)
      if (brokerNodeAndDataMap != null && !brokerNodeAndDataMap.isEmpty) brokerNodeAndDataMap.values.forEach((childData: ChildData) => {
        try {
          val brokerData = childData.getData
          val data = new String(brokerData)
          logger.info("ChildPath: " + childData.getPath)
          logger.error("Broker data: " + data)
          parseAndPopulateMap(map, data)
        } catch {
          case ex: Exception =>
            logger.error("Cluster IP Exception ex: " + ex)
        }
      })

    }
    // return map.entrySet.stream.collect(Collectors.toMap(util.Map.Entry.getKey, (e: util.Map.Entry[String, util.List[String]]) => Joiner.on(",").join(e.getValue)))
    // alternative for above
    val resultMap = new mutable.HashMap[String, String]()
    for ((key, value) <- map)
      resultMap.put(key, value.toArray.mkString(","))
    resultMap
  }
}
